near-sdk/src/json_types/public_key.rs

```diff
@@ -97,31 +132,34 @@ impl<'de> serde::Deserialize<'de> for Base58PublicKey {
     where
         D: serde::Deserializer<'de>,
     {
-        let s: String = Deserialize::deserialize(deserializer)?;
+        let s: String = serde::Deserialize::deserialize(deserializer)?;
         s.parse::<Base58PublicKey>().map_err(serde::de::Error::custom)
     }
 }
 
 impl From<&Base58PublicKey> for String {
     fn from(str_public_key: &Base58PublicKey) -> Self {
-        match str_public_key.0[0] {
-            0 => "ed25519:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
-            1 => "secp256k1:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
-            _ => panic!("Unexpected curve"),
+        match str_public_key.curve {
+            CurveType::ED25519 => {
+                ["ed25519:", &bs58::encode(&str_public_key.data).into_string()].concat()
+            }
+            CurveType::SECP256K1 => {
+                ["secp256k1:", &bs58::encode(&str_public_key.data).into_string()].concat()
+            }
         }
```

Comment:
That's much better that what was here before, but it still creates a bunch of intermediate strings. As a rule of thumb, all good APIs in Rust are organized in a way that allows for avoiding intermediate allocations, so usually there's a way to do only a single alloc. For this case, this would be something like

```rust
let mut buf = String::with_capacity(cap);
buf.push_str("ed25519");
bs58::encode(&str_public_key.data).into(&mut buf);
```

Reply:
The tricky part about this is that we need to know the capacity ahead of time. Digging into the internals of `bs58` crate it does resize and truncate the buffer if it's not exactly that capacity, but it does use some upper bound we can make use of to avoid reallocating. We just gotta be careful if that upper bound ever gets changed from under us whenever we go to upgrade `bs58`.

Also, trying to implement this right now but the API is kinda unwieldy since `.into(&mut buf)` overwrites the initial `push_str("ed25519")`. Grabbing a slice to allocate into without a known capacity ahead of time makes it look pretty ugly. @matklad Would you happen to know of a better way than this:
```Rust
let mut buf = vec![0u8; 8 + upper_bound];
buf.splice(..8, "ed25519:".as_bytes().iter().cloned());
let size = bs58::encode(&str_public_key.data).into(&mut buf[8..]).unwrap();
buf.truncate(8 + size);
String::from_utf8(buf).unwrap()
```

Reply:
Oh wow, looks like I was confused by b58 API and confused you as well @ChaoticTempest , sorry for that!

>As a rule of thumb, all good APIs in Rust are organized in a way that allows for avoiding intermediate allocations

Seems like base58 API isn't setup to do that in a nice way, see https://github.com/mycorrhiza/bs58-rs/issues/77.

Calculating the lengths manually and truncating *can* work, but the complexity is not worth it (and we have to pay extra for duplicate utf8 validation), I suggest sticking with the original impl then!

Reply:
Ahh, no worries. Best to keep it simple then